### ISSUE TITLE
Upgrading to new CUDA API (v2)

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#64]
[creating a separate issue for this.  I expect this issue will be up for a while].

The new CUDA "v2" API has been supported since v4.0 of the CUDA toolkit (i.e. for quite a while; we are now on 7), and probably makes sense to upgrade the Kaldi project to use the new CUDA API.  The old API is deprecated by NVidia, and does not support certain useful functionality such as batched matrix-multiply.

@jtrmal, could you please create a new branch in the kaldi-asr repo called cuda-v2, as a copy of trunk?
@vesis84, it would be good if you could help with this, as you wrote the original cuda matrix library.  The new CUDA API supports having multiple execution contexts.  The current  CuDevice class would be very limiting.  Do you have time to think about the best way to rework things and come up with some kind of sketch of a replacement?  Or if you decide that leaving it mostly unchanged is the
best way, let us know why.

**Comment by jtrmal** [comment]
ok, created. y.

**Comment by danpovey** [comment]
@vesis84, can we assign this issue to you for the time being?  Does not
have to be done super quickly e.g. in the next week or two, but it would be
nice to know that you are on it.
Dan

**Comment by vesis84** [comment]
Hi, I think what you refer to is the support  of 
A) 'concurrent streams', like described here : http://on-demand.gputechconf.com/gtc-express/2011/presentations/StreamsAndConcurrencyWebinar.pdf
(parallel execution of several kernels witing  single GPU)

Or do you mean 'B)' support for multiple GPUs, where each has its own cuda 'cuda context' which is attributed to a separate CPU thread? (Note that separate 'cuda context' means that the contexts don't see the memory of other context unless, we use 'GPU-direct', where I am not sure how the API looks like. Perheps it may be done automatically, via unified address space.)

Case A) can lead to performance improvments only if we have several kernel calls which can be done in parallel in arbitrary order. First, if we have such kernels we can often think how to replace them by single kernel call. Second, such situations do not occur in 'component-based' NN implementation, where the following component is waiting for output  of preceding component.
And third, we should be cautious about using cuda_device_synchronize(), as this would lead to serialization of the run. cuda_device_synchronize() is important to measure accurate runtimes of the functions, being integrated in the CUDA_SAFE_CALL() macro. 
So perhaps there could be some 'wrapper class', which  spawns a vector of kernels, maybe it could be used in convolutional code. This I have implemented as a for loop of matrix-matrix multiplications, where one of the matrices is always 'thin and long'. Then it is a question how to design the interface to support arbitrary function. Mabe something like we have for multi-threading. 

But shortly all this is probably needed for a very specialized cases, while I like about the current code that it is clear and simple. Experimenting with the parallel execution A) would definitely take some time and it is not clear how big would be the benefits.

Just thinking of what is more important, the A) or B)?

**Comment by danpovey** [comment]
I wasn't sure what I meant.
If you are saying that support for multiple concurrent streams and/or
multiple GPUs is not that important, then it would be fine to leave the
interfaces basically as they are (i.e. CuDevice()).  I'm not in great need
of multi-GPU support, because the current parallelization in nnet2/nnet3
works fine for me.

However, there is still the issue of upgrading to the CUDA v2 API.  This
supports things like batched matrix multiply (which I'd like to use in the
convolutional components), and anyway the old API is deprecated.  Do you
agree that upgrading to the v2 API makes sense, independently of any major
changes in the CuDevice interface?  And might you have some time to help
with this at some point- not necessarily all the work, but at least to
start it, and to create a TODO list that others can help with,

Dan


Hi, I think what you refer to is the support of

**Comment by vdp** [comment]
I'm moderately ignorant about the CUDA ecosystem, and failed to find anything about this CUDA v2...
Dan, perhaps you mean cuBLAS API v2?

**Comment by danpovey** [comment]
Yes, that's what I mean.
Dan

**Comment by vesis84** [comment]
Hi, I got confused, I was not sure what was mean by 'CUDA v2 API'. Now I understand that the task is to look into the extended cuBLAS API 2.0 and integrate some new cool features into cudamatrix. Okay, I undestand that this can be helpful. I'll look into the docs, what's in and how different is it from what we have (http://docs.nvidia.com/cuda/cublas/#axzz3iM9bNhe0).

**Comment by danpovey** [comment]
Sure.
It's not actually even necessary to go ahead an integrate any the new
features into the cudamatrix library just yet; once we have switched over
to the new API we can do them one by one as needed.
Dan

**Comment by vesis84** [comment]
Hi, I after reading the intro of the "CUBLAS_Library.pdf", here is the digest of the crucial parts:

Cite:
In general, new applications should not use the legacy cuBLAS API, and existing existing applications should convert to using the new API if it requires sophisticated and optimal stream parallelism or if it calls cuBLAS routines concurrently from multiple threads.

* It looks we are good, unless we want to try simultaneous gemm() via streams, or use multi-GPU setup.
  * Is there some 'cumatrix' function/method running a for-loop of gemm()? 
  * This will be good to play with in a 'unit test' with the typical dimensions.

* the "cublas_v2.h" is very flexible, supporting scenarios:
  * single-thread multi-GPU, 
  * multi-thread single-GPU, 
  * multi-thread multi-GPU scenarios. 
  * We definitely need to do the transition if we decide to use multi-GPUs. For now I'd suggest to add 'cublas-handle' into 'CuDevice', and modify the cublas calls. This is easy to do.

* Other mostly unimportant features / changes:
  * calling cublas from a cuda-kernel (probably not needed) (2.1.9. - min GPU capability 3.5)
  * returning / passing scalars via GPU-memory reference (probably not needed)
  * change in error handling, memory allocation (unimportant)
  * CUBLASXT : courtesy host-side wrapper, hiding memory allocations and transfers (not for us)

**Comment by danpovey** [comment]
I wasn't planning to call simultaneous gemm() via streams.  A couple of
points on this:
  - Firstly, if you want to call simultaneous gemm()s, it's not clear to me
that separate streams are necessary.  Couldn't you just call one, not do
syncthreads, and call call the other one, then do synchtreads()?  I would
have thought separate streams are only necessary if you want to call cuBLAS
from separate threads.
  - Anyway, that's not what I was planning to do.  cuBLAS has a function
that does simultaneous "batched" gemms for you, it's just a regular
function call from the point of view of the user.  It would be nice to use
that (and similar related "batched" operations).
 - The convolutional components would benefit from batched GEMMs.  Of
course it's not urgent.

Dan

**Comment by vesis84** [comment]
So very shortly, the task looks like this:
* switch the API header file to "cublas_v2.h"
* add 'cublas handle' variable into CuDevice, add handle argument to the cublas,
* in a function doing 'tensor' multiplication (for-loop of gemm() calls), try to use streams:
  * functions cudaStreamCreate(); cublasSetStream(); do_gemm(); do_synchronize(); cudaStreamDestroy()
  * measure the acceleration in a unit-test, ask about typical input dimension of the matrices.

**Comment by danpovey** [comment]
Agreed with the first two:

   - switch the API header file to "cublas_v2.h"
   - add 'cublas handle' variable into CuDevice, add handle argument to the
   cublas,

For the last one, regarding tensor  multiplication, I'm not convinced that
it's necessary.  For the case we care about the matrices are all the same
size, and the batched-GEMM function call is sufficient.  But creating a
wrapper for this that takes matrix pointers would make sense (I can do that
later though).
Of course doing that last thing wouldn't hurt just to see how it works.
I was thinking of introducing a way to cancel the synchtreads() after an
operation, but this can wait.

Do you have time to get started with the header changes, or should we fine
someone else?
Dan

**Comment by vesis84** [comment]
I will have time for this in september, for now I should focus on presentation, then I'll be travelling around US. It would be better to find someone to do it.

**Comment by vesis84** [comment]
cite:
  - Firstly, if you want to call simultaneous gemm()s, it's not clear to me
that separate streams are necessary.  Couldn't you just call one, not do
syncthreads, and call call the other one, then do synchtreads()?  I would
have thought separate streams are only necessary if you want to call cuBLAS
from separate threads.

Karel:
* what you propose is mapped into a sequential call of gemms which are not synced with the host
* the streams lead to real parallelism on the GPU, they allow to fill the 'idle' time coming from delay on GPU-memory r/w (the GPU memory has a high throughput and significant delay, which is "hidden" when calling many-thread kernels, as the units are computing while the data are being fetched for oncoming threads)

**Comment by vesis84** [comment]
In the doc is written that "gemm() with streams" is faster than cublasgemmBatched() for larger matrices.

**Comment by vesis84** [comment]
There are also interesting "extension" functions like:
* 2.7.1. cublas<t>geam() : matrix addition with support of transpose
  * C = alpha*op(A) + beta*op(B)
* 2.7.2. cublas<t>dgmm() : multiplication with diagonal matrix
  * C = diag(X)*A; C = A*diag(X)

**Comment by danpovey** [comment]
> There are also interesting "extension" functions like:
>
>    - 2.7.1. cublasgeam() : matrix addition with support of transpose
>       - C = alpha*op(A) + beta*op(B)
>    - 2.7.2. cublasdgmm() : multiplication with diagonal matrix
>       - C = diag(X)*A; C = A*diag(X)
>
>
Interesting.  Currently we are using our own kernels for this; when we
switch over the API we could use this one.
Dan

**Comment by danpovey** [comment]
OK, so the current status on this issue is we are waiting for someone to
find time to upgrade current code to the new API, involving storing the
"stream handle" in the CuDevice object.  I'm going to wait a few weeks and
see if anyone volunteers or if Karel finds time, as this isn't super urgent.
Dan

**Comment by danpovey** [comment]
I just realized that there is a potential use for streams.  What if people want to do multi-threaded decoding and use a GPU for nnet evaluation (using a shared model)?  I think that likely wouldn't work correctly without streams.

**Comment by danpovey** [comment]
@freewym will be working on this.

**Comment by danpovey** [comment]
Done- closing issue.
